### PR TITLE
content: Handle thumbnails

### PR DESCRIPTION
**Pull request by rajveermalviya**
| Thumbnails | Placeholders |
| - | - |
| <video src="https://github.com/user-attachments/assets/19413ccd-cc06-42c0-8894-a1058b6c8bac"> | <video src="https://github.com/user-attachments/assets/3c7d1f77-65b3-4851-a9f1-630e497dfbe3"> |

Fixes: #799

### CODE REVIEW

File: lib/model/content.dart [hunk 1]
```diff
@@ -351,26 +351,51 @@ class ImageNodeList extends BlockContentNode {
 }
 
 class ImageNode extends BlockContentNode {
-  const ImageNode({super.debugHtmlNode, required this.srcUrl});
+  const ImageNode({
+    super.debugHtmlNode,
+    required this.srcUrl,
+    required this.thumbnailUrl,
+    required this.loading,
+  });
```
Comment (gnprice) [inline lib/model/content.dart:359]: Let's split this into two constructors:
```suggestion
  const ImageNode({
    super.debugHtmlNode,
    required String this.srcUrl,
    required String this.thumbnailUrl,
  }) : loading = false;
```

and then an `ImageNode.loading` that makes those other fields null.

That way the constructors enforce that there are these two mutually exclusive configurations.

File: lib/model/content.dart [hunk 2]
```diff
@@ -1028,21 +1053,46 @@ class _ZulipContentParser {
...
       return UnimplementedBlockContentNode(htmlNode: divElement);
     }
 
+    final (linkElement, imgElement) = elements;
+    if (imgElement.className == 'image-loading-placeholder') {
+      return ImageNode(
+        srcUrl: null,
+        thumbnailUrl: null,
+        loading: true,
+        debugHtmlNode: debugHtmlNode);
```
Comment (gnprice) [inline lib/model/content.dart:1070]: ~~so in particular:~~
```suggestion
      return ImageNode.loading(debugHtmlNode: debugHtmlNode);
```

File: lib/model/content.dart (hunk 1 above)
Comment (gnprice) [reply lib/model/content.dart:359]: … Actually, though, scratch that — because I think we want to model this a bit differently.

Even if the thumbnail is still pending, that's no reason to stop the user from opening the lightbox. So let's always have the original URL, which is the one from the `a` element.

File: lib/model/content.dart [hunk 3]
```diff
@@ -1028,21 +1053,46 @@ class _ZulipContentParser {
...
+    }
     final src = imgElement.attributes['src'];
     if (src == null) {
       return UnimplementedBlockContentNode(htmlNode: divElement);
     }
 
-    return ImageNode(srcUrl: src, debugHtmlNode: debugHtmlNode);
+    final String srcUrl;
+    final String? thumbnailUrl;
+    if (src.startsWith('/user_uploads/thumbnail/')) {
```
Comment (gnprice) [inline lib/model/content.dart:1084]: When this condition is false, let's have a check that `src == href`, and return unimplemented if they're unequal.

I believe those were always the same when thumbnailing isn't involved; if there are other cases that exist, I'd like to learn about them, either in manual testing as we use the app or as part of #190.

**Review by gnprice — COMMENTED** [review]
Thanks @rajveermalviya! Taking this straight to integration review because it interacts with the server's 9.0 release timeline.

Here's a partial review before I go AFK for dinner. I've read just the model code so far (lib/model/content.dart).

File: test/model/content_test.dart [hunk 4]
```diff
@@ -459,8 +485,38 @@ class ContentExample {
...
-      ImageNode(srcUrl: 'https://uploads.zulipusercontent.net/8f63bc2632a0e41be3f457d86c077e61b4a03e7e/68747470733a2f2f636861742e7a756c69702e6f72672f757365725f617661746172732f322f7265616c6d2f69636f6e2e706e673f76657273696f6e3d34'),
+      ImageNode(srcUrl: 'https://uploads.zulipusercontent.net/f535ba07f95b99a83aa48e44fd62bbb6c6cf6615/68747470733a2f2f636861742e7a756c69702e6f72672f757365725f617661746172732f322f7265616c6d2f69636f6e2e706e673f76657273696f6e3d33',
+        thumbnailUrl: null, loading: false),
+      ImageNode(srcUrl: 'https://uploads.zulipusercontent.net/8f63bc2632a0e41be3f457d86c077e61b4a03e7e/68747470733a2f2f636861742e7a756c69702e6f72672f757365725f617661746172732f322f7265616c6d2f69636f6e2e706e673f76657273696f6e3d34',
+        thumbnailUrl: null, loading: false),
+    ]),
+  ]);
+
+  static const imageClusterWithThumbnails = ContentExample(
+    'multiple images with thumbnails',
```
Comment (gnprice) [inline test/model/content_test.dart:496]: This can get the same "NoThumbnails" treatment as the single-image case above.

**Review by gnprice — COMMENTED** [review]
Thanks again @rajveermalviya for building this! Done with a review — comments below. Generally it looks great.

The comments are mostly small, but there are a couple that I think might mean glitches a user could spot:
* on _buildAppBarFooter and the height of the app bar
* on addPostFrameCallback vs. scheduleFrameCallback

File: lib/model/content.dart (hunk 3 above)
Comment (rajveermalviya) [reply lib/model/content.dart:1084]: There seems to be multiple cases where `src != href`, forexample src urls starting with `/external_content/` and `https://uploads.zulipusercontent.net/`.
Should they be added one by explicitly (using `startsWith`)?

File: lib/widgets/lightbox.dart [hunk 5]
```diff
@@ -234,17 +252,53 @@ class _ImageLightboxPageState extends State<_ImageLightboxPage> {
 
   @override
   Widget build(BuildContext context) {
+    final thumbnailUrl = widget.thumbnailUrl;
     return _LightboxPageLayout(
       routeEntranceAnimation: widget.routeEntranceAnimation,
       message: widget.message,
+      buildAppBarFooter: _loadingProgress == null ? null : _buildAppBarFooter,
```
[the remark this answers is not in the document]
Comment (rajveermalviya) [reply lib/widgets/lightbox.dart]: It doesn't cause the layout shift, because `_LightboxPageLayout` sets [`Scaffold.extendBodyBehindAppBar`](https://api.flutter.dev/flutter/material/Scaffold/extendBodyBehindAppBar.html) to `true` [here](https://github.com/zulip/zulip-flutter/blob/7285325a3068b40e1feed9db4909aef04d31946f/lib/widgets/lightbox.dart#L189).

File: lib/widgets/content.dart [hunk 6]
```diff
@@ -571,26 +572,32 @@ class MessageImage extends StatelessWidget {
...
           mediaType: MediaType.image));
       },
-      child: resolvedSrc == null ? null : LightboxHero(
-        message: message,
-        src: resolvedSrc,
-        child: RealmContentNetworkImage(
-          resolvedSrc,
-          filterQuality: FilterQuality.medium)));
+      child: node.loading
+        ? const CupertinoActivityIndicator()
```
Comment (rajveermalviya) [inline lib/widgets/content.dart:594]: Also forgot to flag this earlier, the reason for not using the `loader-{black,white}.svg` asset here is that, currently animated icons can't be included in ZulipIcons font (don't know if animated fonts are even possible), I tried and it just displayed a single (first) frame.

So, for now used the `CupertinoActivityIndicator` widget which conveniently is similar to the svg.

**Comment by rajveermalviya** [comment]
Thanks for the review @gnprice, pushed a new revision PTAL :)

File: lib/model/content.dart (hunk 3 above)
Comment (rajveermalviya) [reply lib/model/content.dart:1081]: > they be added one by one explicitly
Did this for now.

File: lib/widgets/lightbox.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (gnprice) [reply lib/widgets/lightbox.dart]: Ah right, that makes sense, thanks.

File: lib/widgets/content.dart (hunk 6 above)
Comment (gnprice) [reply lib/widgets/content.dart:594]: Handy that that's there!

Yeah, I suspect animated fonts are not possible. If we were determined to use the icons from those SVGs, we could find a way (if nothing else, we could rasterize them to high-enough-resolution PNGs). But using this widget instead is a lot simpler and SGTM.

Would you post a screenshot and a video showing what this placeholder looks like? That doesn't require any fiddling on the server — the quick hack I used just now is:
```diff
--- lib/model/content.dart
+++ lib/model/content.dart
@@ -1064,7 +1064,7 @@ class _ZulipContentParser {
     if (href == null) {
       return UnimplementedBlockContentNode(htmlNode: divElement);
     }
-    if (imgElement.className == 'image-loading-placeholder') {
+    if (true || imgElement.className == 'image-loading-placeholder') {
       return ImageNode(
         srcUrl: href,
         thumbnailUrl: null,
```

File: lib/model/content.dart [hunk 7]
```diff
@@ -351,26 +351,49 @@ class ImageNodeList extends BlockContentNode {
...
+  /// This may be a relative URL string. It also may not work without adding
   /// authentication credentials to the request.
   final String srcUrl;
 
+  /// The thumbnail URL of the image.
+  ///
+  /// This may be a relative URL string. It also may not work without adding
+  /// authentication credentials to the request.
+  /// It may be null if Server didn't generate a thumbnail, or doesn't support
+  /// it yet. It will also be null when [loading] is true.
```
Comment (gnprice) [inline lib/model/content.dart:372]:
```suggestion
  ///
  /// This will be null if the server hasn't yet generated a thumbnail,
  /// or is a version that doesn't offer thumbnails.
  /// It will also be null when [loading] is true.
```

(in particular I found the meaning of "yet" a bit hard to parse in the current revision)

File: lib/widgets/lightbox.dart [hunk 8]
```diff
@@ -209,17 +216,30 @@ class _ImageLightboxPage extends StatefulWidget {
...
   @override
   State<_ImageLightboxPage> createState() => _ImageLightboxPageState();
 }
 
 class _ImageLightboxPageState extends State<_ImageLightboxPage> {
+  double? _loadingProgress;
+
+  PreferredSizeWidget? _buildAppBarBottom(BuildContext context) {
+    if (_loadingProgress == null) {
+      return null;
```
Comment (gnprice) [inline lib/widgets/lightbox.dart:236]: Cool, moving this here (from the call site) is a nice simplification too.

File: lib/widgets/lightbox.dart [hunk 9]
```diff
@@ -232,19 +252,54 @@ class _ImageLightboxPageState extends State<_ImageLightboxPage> {
...
+
+    // `loadingProgress` becomes null when Image has finished downloading.
+    final double? progress = loadingProgress?.expectedTotalBytes == null ? null
+      : loadingProgress!.cumulativeBytesLoaded / loadingProgress.expectedTotalBytes!;
+
+    if (progress != _loadingProgress) {
+      _loadingProgress = progress;
+      // This function is called in a build method and setState
+      // can't be called in a build method, so delay it.
+      SchedulerBinding.instance.scheduleFrameCallback((_) { if (mounted) setState(() {}); });
```
Comment (gnprice) [inline lib/widgets/lightbox.dart:277]: nit: lots going on on this line (lots of control flow, with two function literals and a conditional); split it up

In fact I think it benefits from being expanded slightly more, to turn the if-mounted check into an early return to match the pattern it usually appears as:

```suggestion
      SchedulerBinding.instance.scheduleFrameCallback((_) {
        if (!mounted) return;
        setState(() {});
      });
```

File: lib/widgets/lightbox.dart [hunk 10]
```diff
@@ -232,19 +252,54 @@ class _ImageLightboxPageState extends State<_ImageLightboxPage> {
...
+    if (widget.thumbnailUrl == null) return child;
+
+    // `loadingProgress` becomes null when Image has finished downloading.
+    final double? progress = loadingProgress?.expectedTotalBytes == null ? null
+      : loadingProgress!.cumulativeBytesLoaded / loadingProgress.expectedTotalBytes!;
+
+    if (progress != _loadingProgress) {
+      _loadingProgress = progress;
+      // This function is called in a build method and setState
+      // can't be called in a build method, so delay it.
```
Comment (gnprice) [inline lib/widgets/lightbox.dart:276]: Given the discussion in #mobile-team about how these delayed callbacks are generally to be avoided, let's include in the comment an indication of why in this case it's the best available option. That'll help the reader not wonder if we just didn't think it through; if they want to try to find a solution we missed, it'll let them see the information we'd found so they can start from there.

So:
```suggestion
      // The [Image.network] API lets us learn progress information only at
      // its build time.  That's too late for updating the progress indicator,
      // so delay that update to the next frame.  For discussion, see:
      //   https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/addPostFrameCallback/near/1893539
      //   https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/addPostFrameCallback/near/1894124
```

**Review by gnprice — COMMENTED** [review]
Thanks @rajveermalviya for the revision! Just a few nits in this round, below; plus please post a screenshot and video as per https://github.com/zulip/zulip-flutter/pull/820#discussion_r1683626383 above.

File: lib/model/content.dart [hunk 11]
```diff
@@ -1028,21 +1051,51 @@ class _ZulipContentParser {
...
 
-    return ImageNode(srcUrl: src, debugHtmlNode: debugHtmlNode);
+    final String srcUrl;
+    final String? thumbnailUrl;
+    if (src.startsWith('/user_uploads/thumbnail/')) {
+      srcUrl = href;
+      thumbnailUrl = src;
+    } else if (src.startsWith('/external_content/')
+      || src.startsWith('https://uploads.zulipusercontent.net/')) {
+      srcUrl = src;
```
Comment (gnprice) [inline lib/model/content.dart:1086]: nit: indent `if` condition's continuation to not look like part of the body:
```suggestion
    } else if (src.startsWith('/external_content/')
        || src.startsWith('https://uploads.zulipusercontent.net/')) {
      srcUrl = src;
```

File: lib/model/content.dart (hunk 3 above)
Comment (gnprice) [reply lib/model/content.dart:1081]: Yep, that sounds good.

Oh, it looks like we don't have a test case that exercises the `/external_content/` case, though. Do you have an example? That'd be good to add.

**Review by gnprice — COMMENTED** [review]
Oh and two more comments — I'd missed some of the changes in my reading a minute ago. One is a still-smaller nit; the other asks for one more test case.

**Comment by rajveermalviya** [comment]
Thanks for the review @gnprice, pushed a new revision PTAL.

File: lib/widgets/content.dart (hunk 6 above)
Comment (rajveermalviya) [reply lib/widgets/content.dart:594]: Added a video showing placeholders in the PR description.

File: test/model/content_test.dart [hunk 12]
```diff
@@ -420,12 +421,52 @@ class ContentExample {
...
+        '<img class="image-loading-placeholder" src="/static/images/loading/loader-black.svg"></a></div>', [
+    ImageNodeList([
+      ImageNode(srcUrl: '/user_uploads/2/c3/wb9FXk8Ej6qIc28aWKcqUogD/image.jpg',
+        thumbnailUrl: null, loading: true),
+    ]),
+  ]);
+
+  static const imageSingleExternal = ContentExample(
+    'single image external',
+    "https://upload.wikimedia.org/wikipedia/commons/7/78/Verregende_bloem_van_een_Helenium_%27El_Dorado%27._22-07-2023._%28d.j.b%29.jpg",
```
Comment (gnprice) [inline test/model/content_test.dart:463]: Is there an example message for this on chat.zulip.org which you can link to?

**Review by gnprice — COMMENTED** [review]
Thanks for the revision! Everything looks good, modulo one question below.

@alya see the placeholders demonstrated above (and any frame of that video works as a screenshot). They're slightly different from the design we have in web, because this is easiest for reasons discussed at https://github.com/zulip/zulip-flutter/pull/820#discussion_r1682277072 . Given the release timeline (and because this version seems fine) I won't block merging this to wait for feedback on the design, but we can always adjust it later if desired.

File: test/model/content_test.dart (hunk 12 above)
Comment (rajveermalviya) [reply test/model/content_test.dart]: Ha, yeah I'd just noticed I forgot to add that and pushed a new revision with that.

**Comment by gnprice** [comment]
Merged! Thanks again @rajveermalviya for the fast work building this and the swift revisions.
